test/ServiceTest.cpp

```diff
@@ -0,0 +1,179 @@
+  ASSERT_EQ(varToColMap.size(), 2);
+  ASSERT_EQ(varToColMap.at(Variable{"?x"}), 0);
+  ASSERT_EQ(varToColMap.at(Variable{"?y"}), 1);
+  ASSERT_FALSE(serviceOp.knownEmptyResult());
+  ASSERT_EQ(serviceOp.getChildren(), std::vector<QueryExecutionTree*>{});
+}
+
+// Tests that `computeResult` behaves as expected.
+TEST_F(ServiceTest, computeResult) {
+  // Construct a parsed SERVICE clause by hand, see `basicMethods` test above.
+  parsedQuery::Service parsedServiceClause{{Variable{"?x"}, Variable{"?y"}},
+                                           Iri{"<http://localhorst/api>"},
+                                           "PREFIX doof: <http://doof.org>",
+                                           "{ }",
+                                           parsedQuery::GraphPattern{}};
+
+  // This is the (port-normalized) URL and (whitespace-normalized) SPARQL query
+  // we expect.
+  std::string expectedUrl = "http://localhorst:80/api";
+  std::string expectedSparqlQuery =
+      "PREFIX doof: <http://doof.org> SELECT ?x ?y WHERE { }";
+
+  // CHECK 1: Returned TSV is empty -> an exception should be thrown.
+  Service serviceOperation1{
+      testQec, parsedServiceClause,
+      getTsvFunctionFactory(expectedUrl, expectedSparqlQuery, "")};
+  ASSERT_THROW(serviceOperation1.getResult(), ad_utility::AbortException);
+
+  // CHECK 2: Header row of returned TSV is wrong (variables in wrong order) ->
+  // an exception should be thrown.
+  Service serviceOperation2{
+      testQec, parsedServiceClause,
+      getTsvFunctionFactory(
+          expectedUrl, expectedSparqlQuery,
+          "?y\t?x\n<x>\t<y>\n<bla>\t<bli>\n<blu>\t<bla>\n<bli>\t<blu>\n")};
+  ASSERT_THROW(serviceOperation2.getResult(), ad_utility::AbortException);
+
+  // CHECK 3: Returned TSV has correct format matching the query -> check that
+  // the result table returned by the operation corresponds to the contents of
+  // the TSV and its local vocabulary are correct.
+  Service serviceOperation3{
+      testQec, parsedServiceClause,
+      getTsvFunctionFactory(
+          expectedUrl, expectedSparqlQuery,
+          "?x\t?y\n<x>\t<y>\n<bla>\t<bli>\n<blu>\t<bla>\n<bli>\t<blu>\n")};
+  std::shared_ptr<const ResultTable> result = serviceOperation3.getResult();
+
+  // Check that `<x>` and `<y>` were contained in the original vocabulary and
+  // that `<bla>`, `<bli>`, `<blu>` were added to the (initially empty) local
+  // vocabulary.
+  VocabIndex idxX;
+  VocabIndex idxY;
+  EXPECT_TRUE(testQec->getIndex().getVocab().getId("<x>", &idxX));
+  EXPECT_TRUE(testQec->getIndex().getVocab().getId("<y>", &idxY));
+  LocalVocabIndex idxBla = LocalVocabIndex::make(0);
+  LocalVocabIndex idxBli = LocalVocabIndex::make(1);
+  LocalVocabIndex idxBlu = LocalVocabIndex::make(2);
+  EXPECT_EQ(result->localVocab().getWord(idxBla), "<bla>");
+  EXPECT_EQ(result->localVocab().getWord(idxBli), "<bli>");
+  EXPECT_EQ(result->localVocab().getWord(idxBlu), "<blu>");
```

Comment:
Is it (conceptually) important for the `SERVICE` class that these Ids are 0, 1, and 2?
I don't think we want this test to break if the local vocab ever
changes its internal assignment.
The only thing that is important is that it's the id of "bla, bli, and blu".